src/main/java/com/google/sps/servlets/StudyServlet.java

```diff
@@ -0,0 +1,162 @@
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import com.google.gson.Gson;
+
+import com.google.sps.data.Card;
+import com.google.sps.tool.WordSearch;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Comparator;
+
+/**
+ * Handles the study mode questions and responses by adjusting and sorting a card's based on their
+ * familiarity score and last time seen.
+ */
+@WebServlet("/study")
+public class StudyServlet extends HttpServlet {
+  public class QuizCard {
+    private String quizWord;
+    private String cardKey;
+    private ArrayList<String> possibleResponses = new ArrayList<>();
+    private String correctAnswer;
+
+    public QuizCard(Card card) {
+      this.quizWord = card.getTextTranslated();
+      this.cardKey = card.getCardKey();
+      this.correctAnswer = card.getRawText().toLowerCase();
+      this.possibleResponses = WordSearch.generateWordOptions(this.correctAnswer);
+    }
+  }
+
+  private int numOfCardsPerRound = 1;
+  private int maxNumOfRounds = 4;
+
+  // Returns QuizCard's that are sorted by familiarity score as json
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    List<Card> userCards = new ArrayList<>();
+
+    if (userService.isUserLoggedIn()) {
+      String folderKey = request.getParameter("folderKey");
+      Query cardQuery = new Query("Card").setAncestor(KeyFactory.stringToKey(folderKey));
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      PreparedQuery results = datastore.prepare(cardQuery);
+
+      if (results != null) {
+        for (Entity entity : results.asIterable()) {
+          // If the card has no familiarity score then it's set to the default value
+          userCards.add(initializeCard(entity));
+        }
+      }
+      // Sorts cards in increasing order of familiarity score
+      UserCards.sort((Card c1, Card c2) -> Double.compare(c1.getFamiliarityScore(), c2.getFamiliarityScore()));
+      List<List<QuizCard>> quiz = createQuizRounds(userCards);
+
+      Gson gson = new Gson();
+      String jsonResponse = gson.toJson(quiz);
+      response.setContentType("application/json;");
+      response.getWriter().println(jsonResponse);
+    }
+  }
+
+  // Updates and stores a card's new familiarity score
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Entity cardEntity;
+    Double newScore;
+    String answeredCorrectly = request.getParameter("answeredCorrectly");
+    String cardKey = request.getParameter("cardKey");
+    long time = System.currentTimeMillis();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    try {
+      cardEntity = datastore.get(KeyFactory.stringToKey(cardKey));
+    } catch (EntityNotFoundException e) {
+      cardEntity = null;
+    }
```

Comment:
send a response and return after cardEntity = null?

Reply:
You might log this case -- you potentially expected a car to be there and wasn't. By logging details, you can go back later into all your server logs and debug offline..